### PR TITLE
Test for materialized views implementation (CASSANDRA-6477)

### PR DESCRIPTION
**Pull request by aboudreault**
Do not merge, but ready for some review. 

Test for materialized views implementation.

https://issues.apache.org/jira/browse/CASSANDRA-6477

I am waiting on 2 fixes/changes to complete this PR:

* Adding a datacenter or a single node produce unexpected results. (Bad data propagation)
* The drop table and drop column has to be changed to NOT drop the materialized views silently. (see ticket for more info)

### CODE REVIEW

File: materialized_views_test.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+from tools import since, require, new_node
+from assertions import assert_one, assert_invalid
+
+
+@since('3.0')
+@require('6477')
+class TestMaterializedViews(Tester):
+    """
+    Test materialized views implementation.
+    @jira_ticket 6477
```
Comment (ptnapoleon) [inline materialized_views_test.py]: Make that `@jira_ticket CASSANDRA-6477` please

Comment (aboudreault) [inline materialized_views_test.py]: sure. fixed

**Comment by aboudreault** [comment]
@mambocab When you have some time, I would appreciate a first review here :)

File: materialized_views_test.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+from assertions import assert_one, assert_invalid
+
+
+@since('3.0')
+@require('6477')
+class TestMaterializedViews(Tester):
+    """
+    Test materialized views implementation.
+    @jira_ticket CASSANDRA-6477
+    @since 3.0
```
Comment (mambocab) [inline materialized_views_test.py]: Do we have a standard meaning for `@since`?

Comment (ptnapoleon) [inline materialized_views_test.py]: No :(

File: materialized_views_test.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+        session.execute("CREATE TABLE t (id int PRIMARY KEY, v int)")
+
+        for i in xrange(1000):
+            session.execute("INSERT INTO t (id, v) VALUES ({v}, {v})".format(v=i))
+
+        session.execute(("CREATE MATERIALIZED VIEW t_by_v AS SELECT * FROM t PRIMARY KEY (v)"))
+
+        # wait propagation
+        time.sleep(10)
```
Comment (mambocab) [inline materialized_views_test.py]: Are there any logs you can wait on here? If they're distributed logs then keep an eye on https://github.com/pcmanus/ccm/issues/311, which could help you out.

File: materialized_views_test.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+        assert len(result) == 2, "Expecting {} materialized view, got {}".format(2, len(result))
+
+        session.execute("DROP MATERIALIZED VIEW ks.users_by_state;")
+
+        result = session.execute(("SELECT * FROM system.schema_materializedviews "
+                                 "WHERE keyspace_name='ks' AND columnfamily_name='users'"))
+        assert len(result) == 1, "Expecting {} materialized view, got {}".format(1, len(result))
+
+    def drop_column_test(self):
+        """Test that we ... wrong REQUIRE CHANGE IN THE IMPLEMENTATION"""
```
Comment (mambocab) [inline materialized_views_test.py]: What is this docstring about? Same below.

Comment (aboudreault) [inline materialized_views_test.py]: nothing to be done here until the implementation is done

File: materialized_views_test.py (hunk 3 above)
Comment (aboudreault) [inline materialized_views_test.py]: I will recheck that but I don't think I can watch for anything in the log. data are simply replicated async in BG.

File: materialized_views_test.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        session.execute("CREATE TABLE t (id int PRIMARY KEY, v int, v2 int, v3 int)")
+        session.execute("CREATE MATERIALIZED VIEW t_by_v2 AS SELECT * FROM t PRIMARY KEY (v2)")
+
+        for i in xrange(1000):
+            session.execute("INSERT INTO t (id, v, v2, v3) VALUES ({v}, {v}, {v}, {v}) USING TTL 30".format(v=i))
+
+        for i in xrange(1000):
+            assert_one(session, "SELECT * FROM t_by_v2 WHERE v2 = {}".format(i), [i, i, i, i])
+
+        time.sleep(30)
```
Comment (mambocab) [inline materialized_views_test.py]: Oof, this is a long sleep. Anything we can block on?

File: materialized_views_test.py (hunk 3 above)
Comment (mambocab) [inline materialized_views_test.py]: We should probably request that feature. If it'd be useful for us, I imagine it'd be useful for diagnosing problems in the field, no?

File: materialized_views_test.py (hunk 5 above)
Comment (aboudreault) [inline materialized_views_test.py]: Same issue. Not really. I can maybe reduce the the number rows read.... and maybe add a smart_sleep that could wait (30 seconds - *the time of the 1000 assert_one just before*)

**Comment by mambocab** [comment]
I find the use of and relationship between the `prepare` and `prepare_user_table` a little confusing. I think there's sort of already a conventional meaning for the method name `prepare` in the codebase, which `prepare_user_table` subverts. If it were me, I'd add a kwarg (maybe named  `user_table`) to `prepare` and only create that table if the kwarg is truthy.

That's just me being a big ol' nerd, though, and if you disagree don't change anything.

**Comment by aboudreault** [comment]
sounds good, will modify that prepare method.

File: materialized_views_test.py (hunk 5 above)
Comment (aboudreault) [inline materialized_views_test.py]: @tjake Can I use nodetool("replaybatchlog") on all nodes to replace that kind of sleep?

**Comment by aboudreault** [comment]
@mambocab  I pushed more changes since yesterday, your comments/review are welcome.

**Comment by ptnapoleon** [comment]
You use a lot of the python `assert a == b, msg` syntax, rather than the unittest `self.assertEqual(a, b, msg)` which is preferred.

**Comment by aboudreault** [comment]
Ready to merge. 

Big thanks to @ptnapoleon  and @tjake  for their multiple contributions in that big PR.

**Pull request merged**